### ISSUE TITLE
Ideas for improving the No Results block

### ISSUE THREAD

**Issue by annezazu** [WordPress/gutenberg#40240]
In exploring the No Results block as part of the [thirteenth call for testing for the FSE Outreach Program](https://make.wordpress.org/test/2022/03/31/fse-program-testing-call-13-authoring-an-author-template/), a few pieces of feedback have come in to help improve the current experience:

- Update the default text to be grammatically correct/more intuitive. Current: "Add a text or blocks that will display when a query returns no results."
- Consider making the look/feel more like a Group block since it's meant to act as a container.
- Explore how to show that it displays conditionally.
- Consider including it by default in the Query Loop block.

For the last part, the current experience breaks the WYSIWYG feel in many ways since the no results text shows alongside what one would see otherwise. Here are some visuals:

Video of the editor: 

https://user-images.githubusercontent.com/26996883/162858104-6a61a119-e74b-462a-bebd-e21fb91ce0b2.mov

Editor view with no results:
<img width="1331" alt="Screen Shot 2022-04-11 at 6 06 07 PM" src="https://user-images.githubusercontent.com/26996883/162858138-5dc2a697-bd97-4fc9-8f17-c1c6173a4013.png">

Front end view with no results:
<img width="1356" alt="Screen Shot 2022-04-11 at 6 06 13 PM" src="https://user-images.githubusercontent.com/26996883/162858143-f765ae99-8a4c-402a-9e71-7d0e2bba471c.png">

As you can see above, it's hard to tell at a glance how the results will actually look on the front end since the Query Loop is stacked above the No Results block, shifting the positioning.

Justin Tadlock [shared more long form thoughts in his write up here](https://wptavern.com/fse-outreach-13-building-an-author-template-from-the-wordpress-site-editor) that captures pieces of this feedback cohesively: 

>The UI left something to be desired. It simply reads:  Aside from the glaring grammatical error, displaying the message is necessary, but I expected it to look and feel more like the Group block. After all, it is essentially a container for other blocks, and WordPress already has an existing UI for that. It was also odd to visually edit a feature that is conditionally displayed. There should be an indicator that its contents may or may not be shown on the front end. I am not aware of any precedent for such a block in core WordPress, but we should look at other projects and how they handle this. The [Block Visibility](https://wordpress.org/plugins/block-visibility/) plugin by Nick Diego might be a plugin to borrow from. It adds an overlay, border, and icon for contextual blocks.

cc @carolinan since you did so much work to bring this block to life <3

**Comment by carolinan** [comment]
>Update the default text to be grammatically correct/more intuitive. Current: "Add a text or blocks that will display when a query returns no results."
Please suggest an alternative text, so that we can update it before the 6.0 string freeze.

@ndiego Realistically, would you have time to work on a component for displaying blocks conditionally?

For the search results page, we can't rely on the query block knowing if there are results or not while we are in the editor,
so I'm thinking it can't be automated, user action is needed.

WooCommerce relies on the list view and a toolbar option to toggle between empty carts and filled carts:
https://user-images.githubusercontent.com/7422055/162890970-53c0fcbf-c230-4815-8074-a6e80a613f49.mp4

[Block visibility ](https://wordpress.org/plugins/block-visibility/)has a number of settings in the sidebar:

![Block settings sidebar with options to hide the block from all users, some user roles, and on schedule ](https://user-images.githubusercontent.com/7422055/162892096-4accb560-f313-4e81-9e8a-1c7b7d457e20.png)

**Comment by carolinan** [comment]
@justintadlock Can you help me understand what the grammatical error in the text is?

**Comment by justintadlock** [comment]
Sure. "Add a text" in the following should just be "Add text":

> Add a text or blocks that will display when a query returns no results.

**Comment by ndiego** [comment]
@carolinan currently Block Visibility still displays all blocks in the Editor, it only conditionally displays them on the Frontend. I have plans to build some sort of live preview functionality in the Editor, but haven't had a chance to tackle that 😅

**Comment by annezazu** [comment]
Noting that I updated the original text of this issue to include some additional feedback around considering including the No Results block by default in the Query Loop block:

>It would be nice if this was there by default if there is no text.

>I did not know about the “No Results block”. That was interesting to learn about. Perhaps it should automatically be added with the default text if there are no author posts to show

I think including by default will depend on how intuitive it is to know it's shown conditionally and is just a container but wanted to include that piece of feedback for the sake of comprehensiveness in case it comes up in the future.

**Comment by carolinan** [comment]
Are there still any actionable points here to consider for 6.1?

**Comment by annezazu** [comment]
@carolinan the ideas remain above for feedback but I imagine @WordPress/gutenberg-design might have some thoughts around how best to approach this and how to make it more actionable :)

**Comment by jasmussen** [comment]
I created #43168 to address the copy issue, with [this change](https://github.com/WordPress/gutenberg/issues/40240#issuecomment-1101083472) taken into account.

For further improvemetns, here's a question that may hopefully inspire better ideas: can the "No results" block show up in the editor only if there are no results in the query, _or if you select the block in the list view_?

**Comment by jameskoster** [comment]
> can the "No results" block show up in the editor only if there are no results in the query

In some cases we can do this, but we need to support both. If I'm designing my archive template I need to compose both the 'no results' and the post template. It's really very similar to the WooCommerce cart block mentioned above. 

It would be nice to implement a consistent affordance for switching state, as it's something we'll need for other core blocks in the future (login / out, comments loop, comments form). Membership plugins would also rely on such a feature.

A different idea to consider (beyond the scope of this issue, admittedly): since this is something that can affect multiple blocks _at the same time_, we might add options to the view menu, rather than having them local to the block E.g.

> **View**
> * ✓ Logged in
>   * ✓ Admin role
>   * ✓ Gold Member
> * ✖️ Cart empty
> * ✖️ Query empty

This way the orientation can persist across different documents, and you can cater to multiple configurations without having to toggle a bunch of blocks separately.

**Comment by jasmussen** [comment]
I did consider state switching, the way we do that for the HTML block with the preview tab is limited, and use cases like you mention really deserve a better core interface for it.

But in this case, "No results" is not a state of the query block, it's a child block. Which means the way blocks behave at the moment, they'll always show up, no matter the parent state. I guess the query block could have some state handling, but I wouldn't necessarily look there when the feature shows up as its own block. 🤔

**Comment by jameskoster** [comment]
> But in this case, "No results" is not a state of the query block, it's a child block

It really depends on how you look at it haha. I would say it is a state of the Query block – the query has returned nothing. That state effects which child block is displayed – Post Template or No Results. I've always found it strange that the current implementation renders _both_ on the canvas, it totally breaks the wysiwyg.

Another use case for state switching, in case it inspires other ideas – it's similar to how you might want to preview different content in the template you're editing. IE I'm editing my `category` template I might want to see how it presents my 'Recipes', and 'Reviews' categories to get a better sense of the overall impact of my changes.

Maintenance mode is another one.

**Comment by jasmussen** [comment]
> the query has returned nothing [...] That state effects which child block is displayed

Good point. If the query is empty, show the No results block. If it isn't, don't show it at all. But provide a state switching interface (dropdown in inspector?). 

Is there a smaller interim fix that can be applied?

**Comment by jameskoster** [comment]
It's tricky because it's not a setting, so doesn't feel quite right in either the toolbar or the inspector. That's why I think it should be abstracted somehow, but afaik we don't have a mechanism for that either :) 

Perhaps it's worth trying your idea of toggling the visibility of Post Template or No Results based on which block is selected. The initial visibility can be determined by the Query when it's explicit, and default to showing Post Template when it isn't. The trade-off is that this obfuscates the No Results block, but perhaps we can offset that by updating Query to include it by default (currently you have to insert it manually).

**Comment by carlomanf** [comment]
https://github.com/WordPress/gutenberg/discussions/38108 appears to have some relevant ideas for this.
